pkg/k8s/rbac.go

```diff
@@ -0,0 +1,222 @@
+package k8s
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/keikoproj/manager/internal/config/common"
+	"github.com/keikoproj/manager/pkg/log"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+//CreateServiceAccount adds the service account in the target cluster
+func (c *Client) CreateServiceAccount(ctx context.Context, saName string, ns string) error {
+	log := log.Logger(ctx, "pkg.k8s", "rbac", "CreateServiceAccount")
+	serviceAccount := corev1.ServiceAccount{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       common.ServiceAccountKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      saName,
+			Namespace: ns,
+		},
+	}
+	_, err := c.cl.CoreV1().ServiceAccounts(ns).Create(&serviceAccount)
+	if err != nil {
+		if !apierr.IsAlreadyExists(err) {
+			msg := fmt.Sprintf("Failed to create service account %s in namespace %s due to %v", saName, ns, err)
+			log.Error(err, msg)
+			return errors.New(msg)
+		}
+		log.Info("Service account already exists", "serviceAccount", saName, "namespace", ns)
+		return nil
+	}
+	log.Info("Service account got created successfully", "serviceAccount", saName, "namespace", ns)
+	return nil
+}
+
+//DeleteServiceAccount deletes the service account in the target cluster
+func (c *Client) DeleteServiceAccount(ctx context.Context, saName string, ns string) error {
+	log := log.Logger(ctx, "pkg.k8s", "client", "DeleteServiceAccount")
+
+	err := c.cl.CoreV1().ServiceAccounts(ns).Delete(saName, &metav1.DeleteOptions{})
+	if err != nil {
+		if !apierr.IsNotFound(err) {
+			msg := fmt.Sprintf("Failed to delete service account %s in namespace %s due to %v", saName, ns, err)
+			log.Error(err, msg)
+			return errors.New(msg)
+		}
+		log.Info("Service account doesn't exists anymore", "serviceAccount", saName, "namespace", ns)
+		return nil
+	}
+	log.Info("Service account removed successfully", "serviceAccount", saName, "namespace", ns)
+	return nil
+}
+
+//CreateOrUpdateClusterRole create or updates cluster role
+func (c *Client) CreateOrUpdateClusterRole(ctx context.Context, name string) error {
+	log := log.Logger(ctx, "pkg.k8s", "client", "AddServiceAccount")
+	clusterRole := rbacv1.ClusterRole{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: common.RBACApiVersion,
+			Kind:       common.ClusterRoleKind,
+		},
+
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{"*"},
```

Comment:
1. We are creating a cluster role with all the permissions, and associate it with a service account. So the manager will use this service account to access the cluster? 

2. You mentioned users can create a cluster role and SA by themself and provide a SA name, do we need to do some validations? just in case they might make mistake when creating cluster role

Reply:
1. Yes
2. Thats more of a requirement in the documentation and we expect very rare users who are gonna go that route. And, once server and controller implementation is done, we can throw the error as part of the validation where controller tries to verify some validations using service account name and if it fails responds back with that error